### ISSUE TITLE
Make backup storage pluggable

### ISSUE THREAD

**Issue by google-code-exporter** [magro/memcached-session-manager#63]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on October 25, 2009 01:21:36_

Make backup storage pluggable, so that another backend instead of memcached 
might be used, e.g. [<a href="http://ehcache.org/documentation/cache_server.html" rel="nofollow">http://ehcache.org/documentation/cache_server.html</a> &#13;
ehcache server]

_Original issue: http://code.google.com/p/memcached-session-manager/issues/detail?id=19_

**Comment by google-code-exporter** [comment]
_From [gunay.ar...@aerobilet.com](https://code.google.com/u/104131637503070108951/) on July 08, 2012 06:15:41_

or redis server, would be nice.  if you can define the  storage interface , I can work on the redis implementation. 

best regards,
Gunay Arslan

**Comment by google-code-exporter** [comment]
_From [gunay.ar...@gmail.com](https://code.google.com/u/103866699475642162472/) on July 08, 2012 10:17:40_

HI ,

I did try to make a pluggable storage system. the patch is attached. Let me know what you think. 

example for memcahed

 <Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
          memcachedNodes="http://192.168.1.1:8091/pools
          sticky="true"
          sessionBackupAsync="true"
          requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css)$"
          transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory" 
          
          storage.username="sessions"
          storage.password="11111111"
          storage.memcachedProtocol="binary"
          storageFactoryClass="de.javakaffee.web.msm.storage.memcached.MemcachedConnectionFactory" />


example for redis

<Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
          memcachedNodes="http://192.168.1.1:8091/pools
          sticky="true"
          sessionBackupAsync="true"
          requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css)$"
          transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory" 
          
          storage.host="192.168.1.1"
          storage.password="111111"
          storage.database="1"
          storageFactoryClass="de.javakaffee.web.msm.storage.redis.RedisConnectionFactory" />
   </Manager>

**Attachment:** [pluggable.patch](http://code.google.com/p/memcached-session-manager/issues/detail?id=19#hc2)

**Comment by google-code-exporter** [comment]
_From [gunay.ar...@gmail.com](https://code.google.com/u/103866699475642162472/) on July 08, 2012 13:44:58_

attached is fully working copy for memcached and redis. added a proxy storage , so that you can "transfer" sessions from one storage to another , transparently . Sorry for the lack of documentation.

Here is an example configuration  configured redis as master , and memcached as slave. I am trying to move all the sessions from slave to master.

Now I feel better,when I have a chance to contribute back.



<Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
          memcachedNodes="http://192.168.1.1:8091/pools
          sticky="true"
          sessionBackupAsync="true"
          requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css)$"
          transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory" 
          
          storage.masterClass="de.javakaffee.web.msm.storage.redis.RedisConnectionFactory"
          storage.slaveClass="de.javakaffee.web.msm.storage.memcached.MemcachedConnectionFactory"
          
          storage.transferToMaster="true"
          storage.deleteIfFoundOnSlave="false"
		            
          storage.master.host="127.0.0.1"
          storage.master.database="1"
          storage.slave.username="Sessions"
          storage.slave.password="xxxx"
          storage.slave.memcachedProtocol="binary"
          storageFactoryClass="de.javakaffee.web.msm.storage.proxy.ProxyStorageFactory" /> 


best regards,

Gunay Arslan

**Attachment:** [pluggable.patch](http://code.google.com/p/memcached-session-manager/issues/detail?id=19#hc3)

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on July 09, 2012 02:18:35_

Great work, Gunay! I didn't have a look through the complete changeset, do you mind submitting a pull request on github?

Thanx && cheers,
Martin

**Comment by google-code-exporter** [comment]
_From [gunay.ar...@gmail.com](https://code.google.com/u/103866699475642162472/) on July 09, 2012 11:58:27_

Thank you Martin,

what you mean by "pull request " ? What I need to do ? Right now I have a ready only account for the git project.

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on July 09, 2012 12:33:22_

Check out <a href="https://help.github.com/articles/using-pull-requests" rel="nofollow">https://help.github.com/articles/using-pull-requests</a> for a complete description.&#13;
&#13;
TL;DR: fork the repo (from your github account in the web-frontend), make your changes (in a topic-branch) and submit pull-request (also via web-frontend). Then we can discuss changes in github.

**Comment by google-code-exporter** [comment]
_From [gunay.ar...@gmail.com](https://code.google.com/u/103866699475642162472/) on July 09, 2012 13:18:29_

Thank you Martin. 

Did send the pull request.

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on July 10, 2012 06:59:43_

Great, I'll comment in github...

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on July 10, 2012 07:34:53_

Btw, I'm commenting when may day job needs some compile time or anything else. So it may take some time until I'm through :-) I'll let you know when I finished commenting.

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on August 01, 2012 16:57:16_

Sorry for the long delay. Unfortunately I didn't find the time to go through all changes until now.

Looking at the config I have one question: how are dotted attributes like 'storage.memcachedProtocol="binary"' resolved, how does tomcat know where/how to set e.g. memcachedProtocol? Is this documented somewhere?

Some questions regarding attributes, and if they should be moved to storage:

1) What about memcachedNodes (and failoverNodes), shouldn't this be moved to storage.memcachedNodes (with storageFactoryClass="de.javakaffee.web.msm.storage.memcached.MemcachedConnectionFactory")?

2) Regarding redis storage: Jedis seems not to support async commands. This raises the question, what to do with sessionBackupAsync attribute (+ backupThreadCount and sessionBackupTimeout), as it does not make sense for redis.

Btw, from thursday next week I'll be on vacation for 4 weeks, so I will find time to work on this when I'm back.

Cheers,
Martin

**Comment by google-code-exporter** [comment]
_From [downch...@gmail.com](https://code.google.com/u/115424359274750685835/) on April 05, 2013 00:33:47_

Did this patch make it any further?

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on April 05, 2013 00:45:36_

Nope, you can check the status of the related pull request. If you're willing to help you're welcome! :-)

**Comment by google-code-exporter** [comment]
_From [downch...@gmail.com](https://code.google.com/u/115424359274750685835/) on April 10, 2013 04:37:32_

Related pull request: <a href="https://github.com/magro/memcached-session-manager/pull/18" rel="nofollow">https://github.com/magro/memcached-session-manager/pull/18</a>

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on April 10, 2013 05:50:21_

Thanx, I forgot to add the link :-)
